modules/navier_stokes/include/utils/CellCenteredMapFunctor.h

```diff
@@ -82,128 +85,179 @@ class CellCenteredMapFunctor : public Moose::FunctorBase<T>, public Map
+    const auto elem_value = (*this)(elem_arg);
+    // Two term expansion
+    return elem_value + this->gradient(elem_arg) * (fi.faceCentroid() - fi.elemCentroid());
   }
-
-  ValueType evaluate(const ElemSideQpArg &, unsigned int) const override
+  else
   {
-    mooseError("not implemented");
+    mooseAssert(defined_on_neighbor, "We should be defined on one of the sides");
+    const auto neighbor_arg = face.makeNeighbor();
+    const auto neighbor_value = (*this)(neighbor_arg);
+    // Two term expansion
+    return neighbor_value +
+           this->gradient(neighbor_arg) * (fi.faceCentroid() - fi.neighborCentroid());
   }
-};
+}
 
 template <typename T, typename Map>
-std::pair<bool, const Elem *>
-CellCenteredMapFunctor<T, Map>::isExtrapolatedBoundaryFace(const FaceInfo & fi) const
+typename CellCenteredMapFunctor<T, Map>::ValueType
+CellCenteredMapFunctor<T, Map>::evaluate(const ElemQpArg &, unsigned int) const
 {
-  if (!fi.neighborPtr())
-    return std::make_pair(true, &fi.elem());
+  mooseError("not implemented");
+}
 
-  const bool defined_on_elem = _sub_ids.empty() || _sub_ids.count(fi.elem().subdomain_id());
-  const bool defined_on_neighbor = _sub_ids.empty() || _sub_ids.count(fi.neighbor().subdomain_id());
-  const bool extrapolated = (defined_on_elem + defined_on_neighbor) == 1;
+template <typename T, typename Map>
+typename CellCenteredMapFunctor<T, Map>::ValueType
+CellCenteredMapFunctor<T, Map>::evaluate(const ElemSideQpArg &, unsigned int) const
+{
+  mooseError("not implemented");
+}
 
-  mooseAssert(defined_on_elem || defined_on_neighbor,
-              "This shouldn't be called if we aren't defined on either side.");
-  const Elem * const ret_elem = defined_on_elem ? &fi.elem() : fi.neighborPtr();
-  return std::make_pair(extrapolated, ret_elem);
+template <typename T, typename Map>
+typename CellCenteredMapFunctor<T, Map>::GradientType
+CellCenteredMapFunctor<T, Map>::evaluateGradient(const ElemArg & elem_arg, unsigned int) const
+{
+  return Moose::FV::greenGaussGradient(elem_arg, *this, true, _mesh);
+}
+
+template <typename T, typename Map>
+typename CellCenteredMapFunctor<T, Map>::GradientType
+CellCenteredMapFunctor<T, Map>::evaluateGradient(const FaceArg & face, unsigned int) const
+{
+  const auto & fi = *face.fi;
+  const bool defined_on_elem = hasBlocks(&fi.elem());
+  const bool defined_on_neighbor = hasBlocks(fi.neighborPtr());
+  if (defined_on_elem && defined_on_neighbor)
+  {
+    const auto elem_arg = face.makeElem();
+    const auto elem_gradient = this->gradient(elem_arg);
+    const auto neighbor_arg = face.makeNeighbor();
+    const auto linear_interp_gradient =
+        fi.gC() * elem_gradient + (1 - fi.gC()) * this->gradient(neighbor_arg);
+    return linear_interp_gradient +
+           Moose::outer_product(((*this)(neighbor_arg) - (*this)(elem_arg)) / fi.dCNMag() -
+                                    linear_interp_gradient * fi.eCN(),
+                                fi.eCN());
+  }
```

Comment:
we have something similar in GreenGaussGradient.h for faces.
should we use that??

btw should it be interpolated_gradient in that line?
``
    Moose::FV::interpolate(
          Moose::FV::InterpMethod::Average, face_gradient, grad_elem, grad_neighbor, fi, true);

``

rest
```
  if (!is_extrapolated)
  {
    const auto & value_elem = functor(elem_arg);
    const auto & value_neighbor = functor(neighbor_arg);

    // This is the component of the gradient which is parallel to the line connecting
    // the cell centers. Therefore, we can use our second order, central difference
    // scheme to approximate it.
    VectorValue<T> face_gradient = (value_neighbor - value_elem) / fi.dCNMag() * fi.eCN();

    // We only need nonorthogonal correctors in 2+ dimensions
    if (mesh.dimension() > 1)
    {
      // We are using an orthogonal approach for the non-orthogonal correction, for more information
      // see Hrvoje Jasak's PhD Thesis (Imperial College, 1996)
      VectorValue<T> interpolated_gradient;

      // Compute the gradients in the two cells on both sides of the face
      const auto & grad_elem =
          greenGaussGradient(elem_arg, functor, two_term_boundary_expansion, mesh);
      const auto & grad_neighbor =
          greenGaussGradient(neighbor_arg, functor, two_term_boundary_expansion, mesh);

      Moose::FV::interpolate(
          Moose::FV::InterpMethod::Average, face_gradient, grad_elem, grad_neighbor, fi, true);

      face_gradient += interpolated_gradient - (interpolated_gradient * fi.eCN()) * fi.eCN();
    }
```

Reply:
Nice suggestion. Now this routine is only one line, calling `greenGaussGradient`. Hurray for DRY !

Reply:
is there a mistake in greenGaussGradient though?

Reply:
I think it should be
Moose::FV::interpolate(
Moose::FV::InterpMethod::Average, interpolated_gradient, grad_elem, grad_neighbor, fi, true);

Reply:
The only difference is that in GG and MooseVariable FV we only use non-orthogonal correction if we are in 2+D. The correction is zero for 1D so we still ahve correct results, it is just more expensive (not sure if we care about this for 1D problems). I forgot to add this change to CCMF.

Reply:
ugh I must be misunderstanding all this :(
I think it s an error at the code level not on the discretization level

Reply:
This does look like a code bug

Reply:
Apparently only a single test in NS!

Reply:
The reconstructed one ... so fixing this bug in `greenGaussGradient` makes it so I don't have to regold that test relative to when using the custom `CellCenteredMapFunctor::evaluateGradient(FaceArg)` routine. Nice!

Reply:
Thanks Guillaume, that was a great catch! I'm thinking about adding more non-orthogonal test cases in the future where these will definitely pop up.